Run Galaxy fully under uWSGI, including job handlers

natefoo opened the pull request:
## Rationale

Previously, if running Galaxy under uWSGI, it was necessary to configure and start separate Paste or application-only (`scripts/galaxy-main`) Galaxy servers for job handling in order to avoid race conditions with job handler selection, and to separate web workers from job handlers for performance/scalability. This also meant needing some sort of process management (e.g. supervisor) to manage all of these individual processes.

[uWSGI Mules](http://uwsgi-docs.readthedocs.io/en/latest/Mules.html) are processes forked from the uWSGI master process after the application is loaded. Mules can also `exec()` specified arbitrary code, and come with some very nice features:

1. They can receive messages from uWSGI worker processes.
2. They can be grouped together in to "farms" such that messages sent to a farm are received only by mules in that farm
3. They are controlled by the uWSGI master process and can be stopped and started all from a single command line.

## Usage

This PR introduces the ability to run Galaxy job handlers as mules. In the simplest form, you can:

```console
$ GALAXY_UWSGI=1 sh run.sh
```

This will run with a command line like:

```console
$ uwsgi  --virtualenv /home/nate/work/galaxy/.venv --ini-paste config/galaxy.ini --processes 1 --threads 4 --http localhost:8080 --pythonpath lib --master --static-map /static/style=/home/nate/work/galaxy/static/style/blue --static-map /static=/home/nate/work/galaxy/static --paste-logger --die-on-term --enable-threads --py-call-osafterfork
```

You can override these defaults (other than the booleans like `--master` and `--enable-threads`) with a `[uwsgi]` section in `galaxy.ini`, or just configure in `galaxy.ini` and run `uwsgi` directly.

By default, with no `job_conf.xml`, jobs will be run in uWSGI web worker processes, as they were with Paste. This is to keep things simple at first. To run jobs in mules, you only need to start them and add them to the correct farm, which must be named **job-handlers**. Be aware that there are some caveats (below) if you have a `job_conf.xml`. Mules are added in any of the following ways (command line, ini config file, yaml config file):

```sh-session
$ GALAXY_UWSGI=1 sh run.sh --mule=lib/galaxy/main.py --mule=lib/galaxy/main.py --farm=job-handlers:1,2
```
```ini
[uwsgi]
mule = lib/galaxy/main.py
mule = lib/galaxy/main.py
farm = job-handlers:1,2
```
```yaml
uwsgi:
    mule: lib/galaxy/main.py
    mule: lib/galaxy/main.py
    farm: job-handlers:1,2
```

For more handlers, simply add additional `mule` options and add their IDs to the `farm` option.

## Design

Where possible, I have tried to make this as stack-agnostic and purpose-inspecific as possible. There is a bit of ugliness around how mules are designated as job handlers (they have to be in a farm named `job-handlers`), but the goal is to make it easy for anyone going forward to send tasks to mules for asynchronous execution. You'll see a few references to "pools," which is a sort of stack-agnostic abstraction of uWSGI Farms.

For most other functions you might want to push out to mules, it should be as simple as:

1. Add a new message class as in `galaxy.web.stack.message`
2. Create a message handler and register it with `app.application_stack.register_message_handler`
3. Send messages to mules with `app.application_stack.send_message`

For jobs, messages are *only* being used for handler selection. We create `Job`s in the web workers at tool execution time just as we did before, but they are committed to the database with a `null` in the `handler` field, where before they always had to have a handler set at creation time. Mule messages only include the target message handler function, task to perform (`setup`) and job ID of the job. A mule will receive the message and write its `server_name` to the `handler` field, and then pick the job up as they did before without any further jobs code modification.

## Server names

Under uWSGI, server names are manipulated using the template `{server_name}.{pool_name}.{instance_id}` where:

- `{server_name}` is the original `server_name`, configurable in the app config (or with Paste/webless, on the command line with `--server-name`), by default this is `main`
- `{pool_name}` is the worker pool or farm name, for web workers (the processes forked based on the `processes` uWSGI option) this is `web`, for mules this is the farm name, e.g. `job-handlers`
- `{instance_id}` is the 1-based index of the server in its pool, for web workers this is its uWSGI-assigned worker id (an integer starting at 1), for mules this is its 1-indexed position in the `farm` argument

So in the following galaxy.yml:

```
uwsgi:
    processes: 4
    mule: lib/galaxy/main.py
    mule: lib/galaxy/main.py
    mule: lib/galaxy/main.py
    farm: job-handlers:2,3
    farm: something-else:1
galaxy:
    server_name: galaxy
```

uWSGI starts 4 web workers, 2 job handlers, and another mule with `server_name`s:

```
galaxy.web.1
galaxy.web.2
galaxy.web.3
galaxy.web.4
galaxy.job-handlers.1
galaxy.job-handlers.2
galaxy.something-else.1
```

This information is important when you want to statically or dynamically map handlers rather than use the default.

## Caveats

In order to attempt to support existing `job_conf.xml` files that have a default `<handlers>` block, jobs are mapped to handlers in the following manner:

- If you **do not** have a `job_conf.xml`, or have a `job_conf.xml` with **no** `<handlers>` block:
  - If started without a configured `job-handlers` farm or a non-uWSGI server: web workers are job handlers
  - If started with a `job-handlers` farm: mules are job handlers
- If you **have** a `<handlers>` block and **do not** have a `default=` set in `<handlers>`:
  - Works the same as if you have no `job_conf.xml`, **except** explicitly specified static/dynamic handler mappings will result in the specified handler being assigned
- If you have a `<handlers>` block and **do** have a `default=` set in `<handlers>`:
  - The default handler or explicit static/dynamic handler specified is assigned

As before, if a specified handler is assigned and the specified handler is a tag, a handler with that tag is chosen at random. If a handler is assigned due to an explicit static/dynamic mapping, mule messages are not used, the specified handler ID is simply set on the job record in the database.

One way to mix automatic web/mule handling with mapped handling is to define multiple `<handler>`s but not a default, since by default, jobs will be sent to the web/mule handlers, and only tools specifically mapped to handlers will be sent to the named handlers. It is possible to map tools to mule handlers in `job_conf.xml` as well, using `server_name`s `main.job-handlers.1`, `main.job-handlers.2`, ...

This is complicated and perhaps we should do things less magically, but as usually for Galaxy, I am trying to take the approach of least intervention by admins.

There is some functionality included for templating the server name for greater control - e.g. if you run Galaxy on multiple servers, the `server_name` (which is persisted in the database and used for job recovery) would need to include some identifier unique for each host. However, configuration for this is not exposed. In the short term, people in that situation (are there any other than me?) can always continue running handlers externally.

Zerg mode is untested and you would have the potential to encounter race conditions during restarts, especially with respect to job recovery.

## Configurability

I went through multiple iterations on how to make things configurable. For example:

```yaml
---
stack:
  workers:
    - name: default-job-handlers
      purpose: handle_jobs    # this essentially controls the role of the mule, what portions of the application it loads, etc.
      processes: 4
      server_name: "{server_name}.{pool_name}.{process_num}"
    - name: special-job-handlers
      purpose: handle_jobs
      processes: 2
      server_name: "{server_name}.{pool_name}.{process_num}"
    - name: spam-spam-spam
      # default purpose, just run the galaxy application
    - name: eggs
      type: standalone    # "webless" galaxy process started externally
```

This would translate in to a command line like:

```console
$ uwsgi ... --mule=lib/galaxy/main.py --mule=lib/galaxy/main.py \
    --mule=lib/galaxy/main.py --mule=lib/galaxy/main.py \
    --mule=lib/galaxy/main.py --mule=lib/galaxy/main.py \
    --mule=lib/galaxy/main.py
    --farm=default-job-handlers:1,2,3,4 \
    --farm=special-job-handlers:5,6 \
    --farm=spam-spam-spam:7
```

Prior to #3179, I'd made a separate YAML config for the [containers interface](https://github.com/galaxyproject/galaxy/tree/dev/lib/galaxy/containers). These configs use defaults set as class attributes on the container classes, and those defaults are merged recursively down the class inheritance chain.

I wanted to do the same for the stack config, but with #3179, we can start merging YAML configs into the main Galaxy config. Ultimately (after some discussion on the Gitter channel) I've stripped the configurability out until we settle on whether or not and (if yes) how to support the hierarchical configs/defaults in a way compatible with the model @jmchilton created in that excellent PR.

## Invocations

You can start under uWSGI using a variety of methods:

### ini-paste:
```console
$ uwsgi --ini-paste config/galaxy.ini ...
```

### ini
```console
$ uwsgi --ini config/galaxy.ini --module 'galaxy.webapps.galaxy.buildapp:uwsgi_app_factory()' ...
```

### yaml
```console
$ uwsgi --yaml config/galaxy.yml --module 'galaxy.webapps.galaxy.buildapp:uwsgi_app_factory()' ...
```

### separate app config
Galaxy config file (ini or yaml) separate from the uWSGI config file (also ini or yaml):
```console
$ uwsgi --<ini|yaml> config/galaxy.<ini|yml> --set galaxy_config_file=config/galaxy.<ini|yml> ...
```

### no config file
(For example):
```console
$ uwsgi --virtualenv /home/nate/work/galaxy/.venv --http localhost:8192 --die-on-term --enable-threads --py-call-osafterfork --master --processes 2 --threads 4 --pythonpath lib --static-map /static/style=/home/nate/work/galaxy/static/style/blue --static-map /static=/home/nate/work/galaxy/static --module 'galaxy.webapps.galaxy.buildapp:uwsgi_app_factory()' --set galaxy_config_file=config/galaxy.yml --mule=lib/galaxy/main.py --mule=lib/galaxy/main.py --farm=job-handlers:1,2
```

## Logging

By default, everything logs to one stream, and you can't tell which messages come from which process. This isn't bad with one mule, with more it's unmanageable. You can fix this with the following logging config, which includes the use of custom filters in this PR that log the uWSGI worker and mule IDs in the log message:

```ini
[loggers]
keys = root, galaxy

[handlers]
keys = console

[formatters]
keys = generic

[logger_root]
level = INFO
handlers = console

[logger_galaxy]
level = DEBUG
handlers = console
qualname = galaxy
propagate = 0

[handler_console]
class = StreamHandler
args = (sys.stderr,)
level = DEBUG
formatter = generic

[formatter_generic]
format = %(name)s %(levelname)-5.5s %(asctime)s [p:%(process)s,w:%(worker_id)s,m:%(mule_id)s] [%(threadName)s] %(message)s
```

Or, even better, if you switch to a YAML Galaxy application config, you can use a [logging.config.dictConfig](https://docs.python.org/2/library/logging.config.html#logging.config.dictConfig) dict with a special `filename_template` parameter on any `logging.handlers.*FileHandler` handlers that will be templated with `galaxy.util.facts`, e.g.:

```yaml
---
galaxy:
    debug: yes
    use_interactive: yes
    logging:
        version: 1
        root:
            # root logger
            level: INFO
            handlers:
                - console
                - files
        loggers:
            galaxy:
                level: DEBUG
                handlers:
                    - console
                    - files
                qualname: galaxy
                propagate: 0
        handlers:
            console:
                class: logging.StreamHandler
                level: DEBUG
                formatter: generic
                stream: ext://sys.stderr
            files:
                class: logging.FileHandler
                level: DEBUG
                formatter: generic
                filename: galaxy_default.log
                filename_template: galaxy_{pool_name}_{server_id}.log
        formatters:
            generic:
                format: "%(name)s %(levelname)-5.5s %(asctime)s [p:%(process)s,w:%(worker_id)s,m:%(mule_id)s] [%(threadName)s] %(message)s"
```

This will result in 3 log files, `galaxy_web_0.log` (actually contains log messages for all workers), `galaxy_job-handlers_1.log` and `galaxy_job-handlers_2.log`.

## TODO

For this PR:

- [x] Support handler assignment for workflow invocations
- [x] Figure out separate file logging for the workers and each mule
- [x] Make sure `get_uwsgi_args.py` works with a variety of `[uwsgi]` (or lack of) settings in the config
- [x] Finish refactoring and cleaning
- [x] Fix linting, tests, etc., whitelist stricter code guidelines
- [x] Document logging configuration improvements
- [x] ~~Squash WIP commits~~ I think I want to preserve these
- [x] Tests would be great ~~¯\\\_(ツ)_/¯~~ thanks @jmchilton! (ﾉ^_^)ﾉ
- [x] Default to running jobs in worker(s) and don't start mules if there is no galaxy config file and no job_conf.xml and no mules configured.
- [x] Fix thread join failure on shutdown
- [x] Implement correct handler assignment (either web workers or mules) when a default `job_conf.xml` is in place
- [x] Correct server_name templating for mules, ideally: `{server_name}.{pool_name}.{pool_index}` where pool_name is the farm name and pool_index is the mule's index in the farm, or at least `{server_name}.{pool_name}.{server_id}` where server_id is the mule_id
- [x] Correct the information in this PR description
- [x] ~~Also, it shouldn't block merging, but~~ I need to improve the way that a Galaxy determines whether stack messaging should be used for handler selection, and whether a mule is a handler.

## Future TODO

For a post-PR issue:

- Include a default config for proxying GIEs (work started in #2385)
- If possible, mules should probably give up the message lock after some timeout. This may only be possible with signals since `uwsgi.get_farm_msg()` does not have a timeout param.
- Default to uWSGI
- Document recommended job config changes
- Add configurability as described above
- Add configurability for workflow scheduling handlers
- Support Zerg mode
- ~~`get_uwsgi_args.py` won't play nice with `mule` and `farm` settings in the uWSGI config file (`[uwsgi]` section in `galaxy.ini` or wherever it is in your case)~~
- Run handlers without `exec()`: unbit/uwsgi#1608
- Run multiple handler pools for mapping jobs in different ways
- To my knowledge, no other WSGI application stacks support fork/execing asynchronous workers, let alone messaging them. However, I would like to incorporate non-uWSGI worker messaging into the stack code so we could at least send messages to webless Galaxy processes. My initial thought on doing this is to add a stack transport (see `galaxy.web.stack.transport`) that interfaces with the AMQP support already in `galaxy.queues`. But alternatively, maybe the stack messaging stuff should be decoupled from the stack entirely and merged directly in to `galaxy.queues`.

natefoo committed:
Run job handlers as uWSGI mules

natefoo committed:
Script reusage and generate uWSGI mule args

natefoo committed:
WIP job handlers as uWSGI mules from GCC 2017 Hackathon.

natefoo committed:
WIP uWSGI mule job handlers - it works!

natefoo committed:
WIP job handlers as uWSGI mules refactor

natefoo committed:
WIP: Job handlers as UWSGI mules, attempt at unprogrammed mules.

natefoo committed:
Abandoning unprogrammed mules (for now), fixes for programmed mules

natefoo committed:
WIP: attempted to support workers shared between pools, but locking
won't allow for this, since you can't wait on messages from a specific
farm.

natefoo committed:
WIP: drop stack configurability, this will come later. Cleanup for WIP
PR

natefoo committed:
Remove unneeded server_name manipulation from galaxy-main script

jmchilton commented on lib/galaxy/web/stack/__init__.py:
```diff
@@ -14,59 +17,243 @@
+            'process_num': 1,
+            'id': 1,
+            'hostname': socket.gethostname().split('.', 1)[0],
+            'fqdn': socket.getfqdn(),
```
Should this be configurable? Should we synchronize this with galaxy_infrastructure_url configuration in some way? I'm not sure the context so happy to accept no as an answer.

natefoo replied:
In general, yes, this should be more configurable. `server_name` manipulation is not really exposed though at the moment due to the configurability stuff. And yeah, I don't like this hardcoded set of template vars.

Not sure what you mean with synchronizing with `galaxy_infrastructure_url` though.

natefoo replied:
Ah, you meant the FQDN specifically. I think I'd rather leave `fqdn` as it is and add an `infrastructure_hostname` that would be parsed out of the infrastructure URL. FQDN is going to be more useful in cases like mine where you are running one Galaxy instance across multiple hosts. The infrastructure hostname is going to be the same on all of those.

nsoranzo commented:
@natefoo What about the workflow schedulers which can presently be configured with `config/workflow_schedulers_conf.xml` ?

natefoo commented:
@nsoranzo Good thinking, shouldn't be too hard to incorporate.

natefoo committed:
Support workflow scheduling via stack messaging.

natefoo commented:
@nsoranzo, ok handler selection for workflow scheduling is added in c5bd3d4. It's not perfect (it uses the same pool as job handlers) but that'll be addressed when I deal with the configurability items on the TODO list.

nsoranzo commented:
Thanks @natefoo!

natefoo committed:
Start job runner worker threads post fork.

natefoo committed:
Generalize the collection of system/application "facts" usable for string templating.

natefoo commented:
Separating the uWSGI master/worker logs from the mules is proving to be fairly difficult.

I have been working on moving the server_name formatting dict to a more generalized version that could be reused by other components, with the idea that I could also use it to template filenames in the log handler configuration, since I am not aware of any supported way to dynamically select log files for different application instances running from the same [logging.config.fileConfig](https://docs.python.org/2/library/logging.config.html#logging.config.fileConfig) file (i.e. a single `galaxy.ini`).

Having done that, I am not sure that it'll be possible to template the log filename using `fileConfig`. Mules are started with `lib/galaxy/main.py` so it may be possible to monkeypatch or wrap `fileConfig` there to provide an alternate `args` (say, a `template_args` option under `[handler_foo]`) to a handler config that templates the filename beforehand. This would be pretty hacky, though. And the same wouldn't be possible with uWSGI's paste logging support since it imports `paste.script.util.logging_conf` directly in C and sets it up before we have any control in Python.

Using [logging.config.dictConfig](https://docs.python.org/2/library/logging.config.html#logging.config.dictConfig) should give us more options since we could easily manipulate the data structure prior to logging configuration (which would always be performed internally, instead of through PasteScript), but wouldn't be supported with the INI config. However, we could just note that split file logging is not supported unless you switch to the YAML Galaxy app config.

The alternative is [uWSGI log routing](https://uwsgi-docs.readthedocs.io/en/latest/Logging.html#log-routing), except that uWSGI does not, by default, build with PCRE support, and thus does not have log routing capability. We could probably fix this in our wheel, but dividing up the log messages is done with regex, which is less than ideal, and it'd be dependent on the configured log format.

Anyone know of any other solutions?

natefoo committed:
WIP: support log configuration from yaml/json app config with
logging.config.dictConfig. Still working on yaml/json config loading in
the webless app.

jmchilton commented:
There is a binary file in there - ``lib/galaxy/web/framework/.webapp.py.swp`` that I don't think belongs there.

Update: Actually a couple different such files - added in 84f5a5b561f44fe80efe2cb3cfb22dfeb32c7cb3.

natefoo commented:
@jmchilton Thanks, fixed.

natefoo committed:
Finish dictConfig logging feature

natefoo committed:
Fix mule starting under uWSGI with --ini-paste

natefoo committed:
Improve automatic uWSGI startup options interaction with [uwsgi] section
in config file

natefoo committed:
Cleaning, lint and test fixes.

natefoo committed:
Only include --paste-logger in uWSGI args if logging is configured in
the PasteDeploy config

natefoo committed:
Fix for job_handler_count = 0

natefoo commented:
I'm not sure where to document the logging changes until uWSGI becomes the default and/or YAML Galaxy configs become the default. This PR for now, maybe?

Otherwise, I think this is ready for review.

jmchilton commented on run.sh:
```diff
@@ -52,7 +52,11 @@ if [ $INITIALIZE_TOOL_DEPENDENCIES -eq 1 ]; then
 
-if [ -n "$GALAXY_RUN_ALL" ]; then
+if [ -n "$GALAXY_UWSGI" ]; then
+    uwsgi_args="$(python ./scripts/get_uwsgi_args.py)"
```
Did you review common_startup_functions.sh add this release meant to synchronize launching via uwsgi across reports, tool shed, and Galaxy? If no, can you and then comment on whether it makes sense and if we can move get_uwsgi_args.py into that template and share those functions.

I'm working on writing some tests for this PR this morning, if I can't get those done I'll merge this PR this afternoon.

natefoo replied:
I'm taking a look at this.

natefoo replied:
Related, I wonder if the new `galaxy.script` I added should move to `galaxy.util.properties`. I think the only reason I didn't put it in properties was because of all the 3rd party imports for `galaxy.util`?

natefoo replied:
Maybe because of the globals in `galaxy.script` - which was a way to avoid having to modify the scripts that already use these functions the way they do. Probably could just fix that. But maybe `galaxy.util.script` makes more sense anyway.

jmchilton commented on lib/galaxy/web/framework/webapp.py:
```diff
@@ -961,17 +962,30 @@ def build_native_uwsgi_app(paste_factory, config_section):
+    app_kwds = get_app_kwds(config_section, for_paste_app=True)
+    if not app_kwds:
+        # Probably loaded via --ini-paste - expect paste app
+        return None
```
This seems wrong - uwsgi is dying for me with a ``NoneType is not callable`` error not attached to any particular line or stack trace because of this line. Probably because I'm not setting a config file - I don't think that should really be needed but if it is I think we should die with a clearer error here.

natefoo replied:
This was how it was before, but yeah, in expanding this I had a lot of difficulty getting this right. I'll take another look.

It's possible to run under uWSGI with no config file. Can you send the args you were using to get the error you got?

jmchilton replied:
This was before I was serializing the YAML in the test case - so it was the test case without the ``--yaml`` argument.

natefoo replied:
Yeah, I found it. I'd had it working but broke it before I committed and didn't realize it.

jmchilton committed:
Allow test framework to run uwsgi tests.

jmchilton committed:
Add mules integration test and integration test framework updates to support uwsgi.

jmchilton committed:
Fix connection rejected errors for linux when waiting for uwsgi servers in test framework.

jmchilton committed:
Do a better job cleaning up uwsgi processes for tests.

jmchilton committed:
Fix mules test to not require ``app`` in test's memory.

Just run a simple upload using a DatasetPopulator instead of running a tool test that requires the tool be in memory.

jmchilton committed:
Be a more forceful about cleaning uwsgi during tests.

natefoo committed:
Fix a segfault in uWSGI mules caused by trying to access invalid memory
post-fork. There may be more cases like this that we'll have to handle
in the mule fixup function.

Context
=======

uWSGI mules don't fork()/exec(), they fork() and then (as documented)
run supplied code[1]. In order to read the URL parameter for IGV builds
in the default tool_data_table_conf.xml, Galaxy imports urlopen() from
six.moves.urllib.request. In Python 2.7, this imports the same function
from urllib2. urllib2 uses (and reuses) a module global "opener"
object[2]. This is initialized during the prefork application startup
when used for the IGV data table, and then the same object is used
postfork for the mule to load the same data table.

It was discovered when attempting to run the mule job handler test under
macOS, it does not occur in testing on Linux, nor does it occur on macOS
with the default tool_conf.xml.sample and datatypes_conf.xml.sample.
Whether or not the segfault was encountered seemed to depend on which
tools or datatypes were loaded, but not in any clear pattern. The cause
is presumably due to differences in memory allocation schemes between
macOS and Linux, and perhaps garbage collection differences depending on
different config file sizes? Regardless, it seems to be a classical case
of holding on to references to memory that has been deallocated
elsewhere.

The segfault is specifically on the socket_getaddrinfo function in the
_socket C module that provides functionality for the socket module in
the standard Python lib. socket.getaddrinfo() is called by urlopen() by
way of httplib.

uWSGI mule debugging
--------------------

I ultimately figured out the cause of the segfault with lldb. To make
that experience better, I used a debug build of both Python and uWSGI
(uWSGI was probably less important to do but that process is explained
here anyway). Python with debugging is not ABI compatible with non-debug
Python, so building all of Galaxy's C Extension wheels was necessary. In
many cases, pip will do this for you (using the standard pip invocation
in scripts/common_startup.sh), but not all will be successful. To build
debug Python and debug uWSGI and run Galaxy under them, I used the
following steps:

1. You will need a newer version OpenSSL than the one that ships with
   Sierra, I used the one from Brew:

   CFLAGS=-I/usr/local/opt/openssl/include \
   LDFLAGS=-L/usr/local/opt/openssl/lib \
   MACOSX_DEPLOYMENT_TARGET=10.12 \
   ./configure --prefix=$HOME/pydebug --with-pydebug

   Be sure to set $MACOSX_DEPLOYMENT_TARGET. If you do not, the
   `getaddrinfo` function will utilize locking around `getaddrinfo` that
   should not be used[3].

2. After intsalling Python, `virtualenv -p ~/pydebug .venv` in the
   Galaxy root.

3. In the uWSGI source, copy buildconf/default.ini to
   buildconf/debug.ini, set `debug = true` in debug.ini and then (with
   .venv activated) run:

   UWSGI_PROFILE=buildconf/debug.ini python setup.py install

4. Manually install a few problematic dependencies:

   - Use the Starforge recipe for building pysam[4], which mainly means
     grabbing a tarball of @dannon's fork, fixing the version number,
     and running `python setup.py install`. The ssdist tarball for this
     verison on wheels.galaxyproject.org is broken (files are missing
     from its manifest) and I need to fix that. Dannon is also currently
     working on upstreaming his changes and updating Galaxy for the
     latest version of pysam so this shouldn't be a problem in the
     future.

   - The mercurial sdist tarball on wheels.galaxyproject.org is also
     broken, download the correct version tarball directly from the
     mercurial release distribution site and `python setup.py install`
     it. I need to fix this as well.

   - Install numpy before bx-python (if it's not installed before bx is,
     bx will be missing needed features):

     pip install -i https://wheels.galaxyproject.org --no-binary :all: \
        $(grep -i numpy= requirements.txt)

5. Install remaining requirements, iirc the rest should be ok:

   pip install -i https://wheels.galaxyproject.org --no-binary :all: \
      -r requirements.txt

   --no-binary :all: is necessary because there are numerous macOS
   wheels in PyPI with a `none` SOABI tag, meaning these wheels will
   install but will not work for the debug Python ABI.

When you run Galaxy under the debugger, you won't be able to directly
control the fork()ed mules. Instead, you'll need to fire up a second
debugger to catch the mule[5]:

1. Start up Galaxy under lldb, set a breakpoint on fork Be sure that
   your virtualenv is *NOT* activated before starting lldb, or it will
   fail miserably at loading its own python module.

   % lldb .venv/bin/uwsgi
   (lldb) b fork

2. Run Galaxy, e.g.:

   (lldb) r --virtualenv /Users/nate/galaxy/.venv --module galaxy.webapps.galaxy.buildapp:uwsgi_app_factory() --yaml config/galaxy.yml --processes 1 --threads 4 --http localhost:8080 --pythonpath lib --master --static-map /static/style=/Users/nate/galaxy/static/style/blue --static-map /static=/Users/nate/galaxy/static --die-on-term --enable-threads --py-call-osafterfork --mule=lib/galaxy/main.py --farm=job-handlers:1

3. The first fork() is for the workers, so you'll want to skip this:

   (lldb) n

4. The second fork is for the mule. At this point, start a second lldb
   in another terminal, set a breakpoint on something interesting,  and
   attach it to the process that is about to start:

   % lldb
   (lldb) b socket_getaddrinfo
   (lldb) pro at -n uwsgi -w

5. Continue the fork in the original lldb:

   (lldb) n

6. Continue debugging in the second lldb, which is now attached to the
   mule.

[1]: https://github.com/unbit/uwsgi/blob/5039a4b679735348422638baa7cff7a8116762e3/core/mule.c#L95
[2]: https://github.com/python/cpython/blob/20958e6d91d11a80d6c664ce6346791259b1d193/Lib/urllib2.py#L130
[3]: https://emptysqua.re/blog/getaddrinfo-deadlock/
[4]: https://github.com/galaxyproject/starforge/blob/f1013ddf6119381ffba6df06e8d996cee5678280/wheels/build/wheels.yml#L49
[5]: https://stackoverflow.com/questions/14746315/how-to-attach-to-child-process-in-lldb

natefoo committed:
Use fancy uWSGI log messages when running tests under uWSGI, or you can
set whatever log format string you want with `$GALAXY_TEST_LOG_FORMAT`.
If unset, and not running under uWSGI, galaxy.config.logging (the
logging.config.dictConfig dict) will not be set and the default logging
config will be used.

natefoo committed:
Provide failure reason when tool configs fail to load due to system
errors (ENOENT, EPERM, etc.).

natefoo committed:
Improvements for mule job handler testing. Ensure that the server
running the job is a mule. Added a test tool that will write Galaxy
app.config attributes to the output dataset.

natefoo committed:
- Do all logging configuration other than paste fileConfig logging from
  dictConfig. Default log message format varies by stack.
- Refactor a lot of file extension handling in to galaxy.util.path.
- Simplify and improve config file finding.
- Simplify and improve native app creation.
- Integrate uWSGI argument generation with @jmchilton's improvements to
  the startup scripts.
- Fix starting without a config file under uWSGI

jmchilton commented on lib/galaxy/config.py:
```diff
@@ -879,44 +913,21 @@ def configure_logging(config):
-        root.setLevel(level)
-
-        disable_chatty_loggers = string_as_bool(config.get("auto_configure_logging_disable_chatty", "True"))
-        if disable_chatty_loggers:
```
You don't want to preserve this for the paste users?

natefoo replied:
Yeah, I hadn't considered that for Paste. That said, the root logger is set up with level INFO so it shouldn't be a problem. I just checked and you don't see the thread pool messages unless you turn up the root logger to DEBUG. If you did want to turn up root logging to DEBUG you can always set up a separate logger for paste.httpserver.ThreadPool to suppress them.

 And `routes.middleware` is only used in the tool shed.

jmchilton replied:
@martenson Do you recall why we added this option? https://github.com/galaxyproject/galaxy/pull/2735. I have no clue why I opened this PR now.

martenson replied:
@jmchilton Was the routes middleware/toolshed too chatty?

jmchilton commented:
Okay 09feaf6 is pretty awesome - all the best commits happen in the middle of the night I think -
 thanks for looking into that!

I like this PR but I'm starting to feel like the scope of what it is doing is quite large. I'm quite nervous about including this 17.09. If I had an admin hat, I'd appreciate the value of adding this in a release before uwsgi is made the default - with my dev hat on though I'd like to sit with these changes a lot longer before doing a release and I'd like to swap the config file at the same time.

jmchilton commented:
@natefoo

1. Do you have any intention of using mules in production?
1. If yes, if we delayed this until 18.01 would your setup allow you to test them on test before the release 18.01 or can test and and production not diverge so much?

natefoo commented:
Yeah, although most of the changes only affect running under uWSGI, I suppose there are a number of changes that could potentially break non-uWSGI setups. I hate to delay this further but would you feel comfortable making uWSGI the default in dev-post-17.09 and 18.01 even without a "try it out if you want" release in between? I hate the idea of pushing uWSGI-as-default out until 18.05.

martenson commented:
> would you feel comfortable making uWSGI the default in dev-post-17.09 and 18.01 even without a "try it out if you want" release in between?

yes

natefoo commented:
1. So... the sad thing for me is that I will not be able to use handler mules on Main due to its unique setup: Main is spread out between 4 different VMs, two of which serve web requests (uWSGI) and 2 of which handle jobs ("webless" servers) and Pulsar staging (uWSGI). I may switch the handlers to start as mules under the Pulsar staging uWSGI for convenience, but I won't be able to use messaging for handler assignment since uWSGI can only message mules that it starts and the jobs are created on different VMs.

   I will switch Test to use them, however, since it runs all on one VM. I will need to get Zerg mode working first, though. I have some ideas for this.

   Additionally, I'd like to hook Galaxy's AMQP support into the stack messaging stuff (perhaps a bit of a misnomer in that case) so that job dispatching can be done via AMQP for Main and anyone else who'd prefer it. That wouldn't use mule messaging but would use many of the same code paths.

2. If this is merged in dev post-17.09-branch, I can use it on Test (as soon as I get zerg mode working).

natefoo commented:
Note to self: Probably want to rebuild the wheel with PCRE for routing support, especially as it is probably? needed for the GIE proxying? Not sure about that though. But if so, it should be done before the release so we don't have to do any kind of wheel versioning messiness.

jmchilton commented:
> would you feel comfortable making uWSGI the default in dev-post-17.09 and 18.01 even without a "try it out if you want" release in between?

Not at all - but would I feel more comfortable with one big change versus two medium changes if the big change had more time in development? Perhaps.

The thing I would say to mitigate your worry about pushing the default switch out too far is that if we switch the config from ini to YAML at the same time and dispatch on that to determine if paste or uWSGI should be used by default (the way reports and shed does in this release) - we can make the default for new installations uWSGI in 18.01 and delay the paste elimination for older installations until such a time as it is absolutely needed. This seems a fair compromise between avoiding breakage and pushing the platform forward.

martenson commented:
Will @erasche and @mvdbeek be able to run 18.01 dev?

natefoo commented:
@jmchilton So if I understand you correctly, you're saying that in 18.01 we'd use uWSGI for new installs, and in 18.05 we'd do the automatic ini->yaml conversion? I'd be ok with that.

jmchilton commented:
@natefoo First part yes, new install use uWSGI in 18.01. Second part no, I'd never advocate automatically converting deployer's files (though there is some tooling for a semi-manual approach included in this release). Instead, I'm just saying at some point in the future we would interpret a ini file as also a sign to use uWSGI by default - not necessarily in 18.05 - just at whatever point the features missing in paste out weigh the disadvantages of potentially changing people's default server and breaking middleware or something.

Update: It seems like I am arguing hard for this - I'm not. I'm really ambivalent - if you want this in this release - I'm happy to help with that.

natefoo commented:
Ok, I'm game. :sob:

EDIT: this means I am ok with waiting for 18.01.

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -1228,6 +1228,16 @@ use_interactive = True
 #job_config_file = config/job_conf.xml
 
+# NOTE: This option is temporary until additional configurability for starting
+# and pooling asynchronous workers under uWSGI is added.
```
I'd move this sentence below after `#this config file.`

nsoranzo commented on lib/galaxy/app.py:
```diff
@@ -208,8 +207,13 @@ def postfork_sentry_client():
+
         self.model.engine.dispose()
         self.server_starttime = int(time.time())  # used for cachebusting
+        # When running the application without a web stack, this signals the application loop to break and call the shutdown method
```
This comment is a bit confusing for me, when `self.exit = False` the application loop should *not* break, right?

nsoranzo commented on lib/galaxy/util/properties.py:
```diff
@@ -150,4 +138,31 @@ def before_get(self, parser, section, option, value, defaults):
 
 
+def __get_all_configs(dirs, names):
+    return filter(os.path.exists, starmap(os.path.join, product(dirs, names)))
```
`filter` in Python3 returns an iterator, so you need to add a `list(...)` around it (since `__find_config_files()` expects a list).

nsoranzo commented on lib/galaxy/web/stack/__init__.py:
```diff
@@ -14,59 +15,278 @@
+from six import string_types
+
+from .message import ApplicationStackMessage, ApplicationStackMessageDispatcher
+from .transport import ApplicationStackTransport, UWSGIFarmMessageTransport
```
These 2 local imports should go after the `from galaxy` imports.

natefoo committed:
Attempt uWSGI restarts when running for testing. It's not able to detect
if it failed due to port binding, but it'll retry if uwsgi exits
non-zero.

jmchilton committed:
Rework cleanup of uwsgi in tests...

This looks cleaner and was closer to what I originally wanted - undoes some stuff @natefoo did hopefully he isn't mad.

jmchilton committed:
More mule testing.

natefoo commented:
@jmchilton I AM SO MAD RIGHT NOW nah just kidding. I didn't like my way of trying ports, yours is good too. The only thing mine did that yours doesn't is actually check whether uwsgi ran and failed or not, yours just kills it (regardless of whether it's running) and tries it again. In that case you might want to bump `sleep_tries` back up to the default 150 (and remove the kwarg passing I added since it wouldn't be needed), or else it only gives uWSGI 5 seconds to start.

jmchilton committed:
Don't require integration tests to use the same port for every Galaxy instance.

The previous logic would pick a port at random if one wasn't explicitly defined and then set an environment variable requiring each instance to re-use that. This new logic doesn't reuse randomly assigned ports that way by tracking if a port was randomly assigned or not.

jmchilton committed:
Just skip the mule tests on Jenkins.

They work locally alone or ran together, they work locally in the same Docker container as Jenkins, I don't understand.

jmchilton commented:
Pushed a commit to just skip the tests during CI - they work locally every different way I slice it - even mixing paste and uWSGI tests in the sam container Jenkins is using - it just not work on Jenkins. I don't understand but my lack of understanding shouldn't slow down this PR.

natefoo commented:
I can reproduce it locally. Are you testing on mac or Linux?

jmchilton commented:
@natefoo I was testing on a Mac - but I thought that wouldn't matter with Docker. Can you reproduce it by running just a single or a couple tests - or do you need to run the whole suite?

natefoo commented:
You're running under docker on a mac (so Linux)? It reproduces for me if I run `nosetests test/integration`.

jmchilton commented:
Rebase and fix the linting issues (I can do those if you'd prefer) - or are you waiting for me to figure out the integration tests? I was planning to merge with those just set to skip I think.

natefoo commented:
I'll rebase or merge and fix, sure. I was really hoping to have the test working though, so I was going to re-WIP this PR and circle back to that issue soon.

jmchilton commented:
Tests are good.

natefoo commented:
Especially with a big change like this, and the fact that it only has 1 test at the moment. =)

jmchilton commented:
I'm working on a fix today for my collection state branch that has integration tests like this where when they are run together they fail but work fine separately. I'm going to teach Galaxy about postgres template databases to generate a new postgres database for each test from the existing empty database in the Docker image. The rest of the app is isolated this way from test to test - so the database should be as well.

natefoo commented:
Oooh, very nice, thanks. This is what's causing the error we're seeing here you think?

jmchilton commented:
Okay well with the rocky addition of #4910, #4914, #4893, and #4887 - the integration tests should do a much better job cleaning up after themselves and not interfering with each other. I suspect if you can rebase against dev the tests will work (unless it is a port thing above the Galaxy layer).

natefoo committed:
Revert "Just skip the mule tests on Jenkins."

This reverts commit 32a280bf3100f359e1d046206d47a36e0d92db62.

natefoo committed:
Merge remote-tracking branch 'origin/dev' into job-handler-mules

natefoo committed:
Fix path module divergence

natefoo committed:
Fix post-merge lint issues

natefoo committed:
Fix call to __listify.

jmchilton commented:
Well I guess I was wrong. #4962 might also fix the tests - but I'd just revert the tests again and we can follow up.

natefoo commented:
=(

natefoo committed:
Revert "Revert "Just skip the mule tests on Jenkins.""

This reverts commit d7aec7e04dc064fd9e133d9a592e1cb5ef486037.

jmchilton commented:
Still has failing unit tests and import order linting issues - let me know if you want me to take a crack at the unit tests.

natefoo commented:
Working on them, thanks

natefoo committed:
Fix an iterator bug with find_config_file() when using include_samples
and write a doctest for it.

natefoo committed:
Fix config file finding in the orm scripts utility and make its doctest
clean up after itself

natefoo committed:
Fix import linting errors

natefoo commented:
I should probably move the `galaxy.util.path` import changes to a separate PR.

natefoo committed:
Clarifications/fixes from @nsoranzo's comments

natefoo committed:
Undo changes unrelated to uWSGI mule work

natefoo committed:
Revert script changes unrelated to uWSGI mule work. Includes
galaxy.util.script so that get_uwsgi_args.py still works, but that will
be PR'd separately and merged later.

natefoo committed:
Fix typo in facts module

natefoo commented:
Logging is documented in the comment, and I am working on writing new admin documentation that will include it.

However, attempting to start a fresh clone of this with `GALAXY_UWSGI=1 sh run.sh` now fails. So this is not ready to go.

natefoo committed:
Fix tool migrations with config file changes, do everything from
manage_db.sh.

natefoo committed:
Merge remote-tracking branch 'origin/dev' into job-handler-mules

natefoo committed:
Fix use of configure_logging() by scripts that pass the config as a dict
rather than a galaxy configuration object

natefoo committed:
Merge remote-tracking branch 'origin/dev' into job-handler-mules

natefoo committed:
Reenable automatic silencing of chatty loggers by default. This is
needed when Paste is used but PasteDeploy is not, e.g. with tests.

natefoo committed:
Fix None server_name when using the default under uWSGI

natefoo committed:
Move signal handling from uWSGI stack to main

natefoo committed:
Improve shutdown cleanliness under uWSGI.

natefoo committed:
Fix lint errors

natefoo committed:
Remove deprecated job runner URL config options

natefoo committed:
Do not autostart uWSGI mules

natefoo committed:
Postfork thread start fixes

natefoo committed:
Run jobs in web workers if no job-handlers farm is configured and using
a default job conf (no job_conf.xml)

natefoo committed:
Add multi-farm support to the main script

natefoo commented:
I waffled on the `job_handler_count` option in the Galaxy config, which was only supposed to be temporary anyway, and removed it. I did this because it causes too much confusion with `job_conf.xml`, which is where all job config-related stuff is supposed to go.

Also, by default, we won't start any mules, and web workers will just handle jobs. In the future we might want to default to starting a mule, but for now that makes an already big change even bigger and more likely to encounter issues.

So as of now, if you want mules, you must explicitly specify them on the command line, or in the uWSGI config like so (yes, the clobbering key syntax is correct in all cases):

```sh-session
$ GALAXY_UWSGI=1 sh run.sh --mule=lib/galaxy/main.py --mule=lib/galaxy/main.py --farm=job-handlers:1,2
```
```ini
[uwsgi]
mule = lib/galaxy/main.py
mule = lib/galaxy/main.py
farm = job-handlers:1,2
```
```yaml
uwsgi:
    mule: lib/galaxy/main.py
    mule: lib/galaxy/main.py
    farm: job-handlers:1,2
```

As for how handler mapping works (or should work):

- If you **do not** have a `job-handlers` farm
  - If you have no `job_conf.xml`, the web workers will automatically be used as handlers
  - If you have a `job_conf.xml` and
    - It has no `<handlers>` section, we should use the web workers as handlers (**TODO**)
    - It has a `<handlers>` section but it only has 1 handler and that handler is `main`, we should use the web workers as handlers (**TODO**)
    - It has any other `<handlers>` config, jobs are assigned to handlers as per that config, web workers will not run jobs
- If you **do** have a `job-handlers` farm
  - If you have no `job_conf.xml`, the mules will automatically be used as handlers (and web workers will not be used)
  - If you have a `job_conf.xml` and
    - It has no `<handlers>` section, we should use the mules (**TODO**)
    - It has a `<handlers>` section, jobs are assigned to handlers as per that config, mules will not run jobs unless their server names match a handler id in the job config.

natefoo committed:
Fix/improve server name manipulation

natefoo committed:
Attempt to deal with job_conf.xml handlers in a backwards-compatible
way.

natefoo commented:
I've updated the top comment with the description for the actual implementation of job/handler mapping.

natefoo committed:
Fix unit tests

natefoo commented:
:shipit:

natefoo commented:
This is ready for new reviews.

jmchilton commented on lib/galaxy/config.py:
```diff
@@ -349,7 +342,6 @@ def __init__(self, **kwargs):
         self.smtp_password = kwargs.get('smtp_password', None)
         self.smtp_ssl = kwargs.get('smtp_ssl', None)
         self.track_jobs_in_database = string_as_bool(kwargs.get('track_jobs_in_database', 'True'))
-        self.start_job_runners = listify(kwargs.get('start_job_runners', ''))
```
The config linting framework has the ability to flag such options - we should remember to document these there.

jmchilton commented:
Congrats @natefoo - nice work!

martenson commented:
Splendid! Did we also agree on pronunciation of uwsgi? :)

natefoo commented:
https://www.youtube.com/watch?v=E5POYrUawW8#t=29

natefoo commented:
Thanks for merging!
